### ISSUE TITLE
Volumes randomly failing

### ISSUE THREAD

**Issue by phal0r** [longhorn/longhorn#341]
From time to time containers with Longhorn volumes randomly fail.  We need to redeploy them via Rancher UI to recover from this state.

I can find a lot of entries like this in longhorn-manager (can't tell if it is related to the failing):
`Requeue longhorn-system/rancher-2-node-02 due to conflict`

**Comment by yasker** [comment]
@phal0r this message is only at debug level, and it won't be the reason that pods are failed.

Can you show the manager's log when the incident happened? Also, let us know if you noticed anything else abnormal.

Just in case, make sure all of your nodes are in the same location and have good (high bandwidth and low latency) network connections between them.

**Comment by phal0r** [comment]
Thanks for the quick response. It sounded quite critical, this is why I wanted to post it. 
It's a Rancher 2.1 cluster consisting of 3 nodes in the same data center at digitalocean, so bandwidth and latency should be fine.

Once it happens again, I will try to get the logs.

**Comment by phal0r** [comment]
I could find a Error in the Logs of the manager now:
```
E1128 15:22:29.350458 1 engine_controller.go:542] failed to update engine pvc-98296465-d702-11e8-ad59-4ed6ddea6d45-e-8d6b8fdb to stop monitoring: failed to reste engine status for : Operation cannot be fulfilled on engines.longhorn.rancher.io "pvc-98296465-d702-11e8-ad59-4ed6ddea6d45-e-8d6b8fdb": the object has been modified; please apply your changes to the latest version and try again
```

The volume gets detached and the corresponding containers are failing and reporting "read-only filesystems". Only a manual redeploy of the container will reattach the volume.

**Comment by yasker** [comment]
This error message itself is not critical and isn't the reason for detaching. But it meant the manager decide to detach the volume (thus stop the monitoring). So the volume was started detaching before this message.

Can you show the fuller log of the managers around the same time? Also, you can give a look at the log of <volumename>-e-xxx which is the engine of the volume. If it failed due to some reason, it can result in automatic detaching of the volume.

**Comment by yasker** [comment]
@de-robat 

```
time="2018-11-28T15:22:29Z" level=debug msg="Detaching volume pvc-98296465-d702-11e8-ad59-4ed6ddea6d45 from rancher-2-node-01"
...
time="2018-11-28T15:22:45Z" level=debug msg="Attaching volume pvc-98296465-d702-11e8-ad59-4ed6ddea6d45 to rancher-2-node-03"

```

This means the Longhorn manager decide to detach the volume. Seems immediately after attached the volume to `rancher-2-node-01`, it was detached and reattached to ` rancher-2-node-03`.  Was the pod respun on a different node? It may due to Kubernetes ask it to, e.g. original pod which uses the volume crashed somehow by itself, then Kubernetes decided to start it on another node. Can you show a few lines above this snippet?

Reading about the original message. What failure are we talking about here?

**Comment by de-robat** [comment]
Sure @yasker,
 up until 4 days before the incident the log does not contain anything else than:
```
time="2018-11-24T10:31:06Z" level=debug msg="Requeue longhorn-system/rancher-2-node-02 due to conflict"
```
the next entry of interest may be:
```
time="2018-11-24T09:14:17Z" level=info msg="Longhorn replica longhorn-system/pvc-7f7c12ab-ccdf-11e8-a074-32dca53fbd53-r-ac28e8cf has been deleted"
```
but that does not even relate to the same volume so may not related to our issue neither?

To elaborate a little more on the error-scenarios we do encounter:

We run services, e.g. a next-cloud installation that utilizes a postgres for storage purposes. This postgres is configured to use longhorn as it's storage driver and mostly runs reliable. Then, from time to time, it happens that we can not access the next cloud installation anymore. The logs of the postgres then read something along the lines of "can not write to disk, reason: volume is read only". Thus we have to redeploy the postgres in error and everything starts working again, no special interaction with the longhorn services needed as it seems, that they have already recovered the erroneous state.

We see these kinds of errors not just with next-cloud but with other services that have a longhorn-based persistence such as wordpress and open-ldap. 

So i figure for whatever reason the volume needed to be moved by longhorn (the reason is of course of interest, i just can't see one yet) and after being moved it is considered Read-Only by the services that are using the corresponding volume claim. I neither can't see any server-downtimes for the erroneous timeframes nor any reorganization of the pods. But then again, we mostly encounter the errors when we do need the services and may have restarted them a little to eagerly. So, if it helps, next time i will provide a full picture of the logs of all services involved as well as check if the servers involved had any ongoing maintenances.

**Comment by yasker** [comment]
@de-robat 

Now it's an interesting case.

Longhorn itself won't initiate the moving of volume. Only API caller or Kubernetes can decide to do so. I am not sure the log I saw above is before or after the event. It can be in the effort of redeploying the workload, Kubernetes relocate the pod to another node so Longhorn moved along.

Can you show me the immediate log for the one minute before:
```
time="2018-11-28T15:22:29Z" level=debug msg="Detaching volume pvc-98296465-d702-11e8-ad59-4ed6ddea6d45 from rancher-2-node-01"
```

Also, the next time when this happened, can you collect all the logs from different related components before redeploying? See https://github.com/rancher/longhorn/blob/master/docs/troubleshooting.md#troubleshooting-guide The logs from engine and replicas are critical and they will be gone after redeploying. Logs from CSI plugin are helpful to decide if Kubernetes decides to move the volume.

There is only one situation that Longhorn manager will decide to detach the volume without API calls, and it will due to the Longhorn engine failure. Though here it's reattached rather than just detach, and if the engine fails, the managers will collect the engine log automatically (I think there are more than one manager in your case? all logs are helpful), which I didn't see here as well.

BTW, if there are too many logs, you can pack them and send it directly to me, by e.g. email or slack.

**Comment by de-robat** [comment]
@yasker, unfortunately there is no immediate log before the event at least not in the managers log. The log looks like follow:
```
…
time="2018-11-28T10:49:24Z" level=debug msg="Requeue longhorn-system/rancher-2-node-02 due to conflict"
time="2018-11-28T11:15:39Z" level=debug msg="Requeue longhorn-system/rancher-2-node-02 due to conflict"
time="2018-11-28T11:15:39Z" level=debug msg="Requeue longhorn-system/rancher-2-node-02 due to conflict"
time="2018-11-28T13:10:39Z" level=debug msg="Requeue longhorn-system/rancher-2-node-02 due to conflict"
time="2018-11-28T15:22:29Z" level=debug msg="Detaching volume pvc-98296465-d702-11e8-ad59-4ed6ddea6d45 from rancher-2-node-01"
time="2018-11-28T15:22:29Z" level=debug msg="Start monitoring pvc-98296465-d702-11e8-ad59-4ed6ddea6d45-e-8d6b8fdb"
…
```
I think kubetail will not offer anything new, as the issue mostly is resolved by the longhorn system when we do restart our pods. So I won’t expect seeing anything there, but I prepared the cluster to give it a try. On a site note: Your troubleshooting guide says `kubetail longhorn-system -n longhorn-system` is this information outdated? Most of the pods are not named longhorn-system-*, so I went with `kubetail longhorn -n longhorn-system` instead.
I manually scraped trough the logs of the csi-plugin and tried corelating the time of the managers log with logs in the csi-plugin, but no luck. There is no log after the 23d of November.
And yes, the setup is comprised of 3 managers as seen here. 

![image](https://user-images.githubusercontent.com/7009867/49576016-6b69fa80-f944-11e8-879f-2f55c309a32a.png)

I’ll pack the logs of all the involved components for you, as I’m not quite sure where to look and will provide them via mail.

**Comment by yasker** [comment]
@de-robat Sorry there is a typo in the troubleshooting guide. It's meant to be `kubetail longhorn-manager -n longhorn-system`. It's fixed now.

**Comment by de-robat** [comment]
Well, no worries @yasker , it was still comprehensible, was just wondering wheter it strives for everything in the longhorn-namespace or just the managers. Good thing you clarified it though 👍 btt: As requested by you within our slack converastion i'm going to provide more logs then the initial set of logs from yesterday to give you the fullest picture possible. And thanks again for all your effort so far!

**Comment by yasker** [comment]
Hi Robert, I’ve seen the log. For now, it looks like something is killing Longhorn engines and replicas. If you take a look at logs for rancher2-node3 and rancher2-node1, you will notice the report of `instance xxxx crashed`, which means it’s a stop not initiated by Longhorn system, which should the reason of why the volume is gone. The manager has logged the final logs of the engine. And the cases I saw so far, engine all reports `Receive terminated to exit`, which means SIGTERM or SIGINT was sent to the process to shut it down. I suspect Kubernetes or Linux kernel is trying to do this. Can you check the Kubernetes log to see if it’s doing pod eviction, or kernel is using OOM killer? If your cpu/memory is very tight on the node (seems specially on rancher2-node3), it can happen.

**Comment by de-robat** [comment]
I tried forcefully shutting down one of our nodes that holds a volume which is in use, but no luck regarding forcing our issue ( which in it self is a good thing :D )
In parallel i kept running kubetail which did not reveal any issues neither. So it will take me some more time to provoke the erroneous scenario. As soon as i can provide more details i'll come back to you! Thank you so much for your time and effort!

**Comment by yasker** [comment]
@de-robat No problem. Let me know if you found anything. Maybe it's due to a large workload has been deployed since it's likely caused by high CPU/memory utilization on the node. And if it happens again, kubelet log and dmesg should have some clues.

**Comment by phal0r** [comment]
We did some more testing and could reproduce a scenario, where volumes would be read-only or not available at all. This happens, when we restart the docker service of the node, where more longhorn components are running: Specifically csi-attacher, csi-provisioner, driver-deployer and longhorn-ui.

All Volumes on this node get killed obviously, when restarting the service, but do not recover. We suspect, that some of these components need to be made high available also. Would you agree or is this just by coincidence?

We also gathered more logs and @de-robat will send you the link to these logfiles. Hopefully the logs give more insights on why recovery fails.

**Comment by yasker** [comment]
@phal0r Thanks for the followup.

We didn't take the scenario that Docker service crashed but no node rebooted into consideration. If the node was rebooted, the manager will detect that and decide it's normal and we should reattach the volume. Though if only the docker service restarted, the manager still decides something was wrong and then detach the volume.

We didn't automatically reattach the volume is because in some cases if volume failed to start during the starting period, we're going to introduce an endless loop of attaching volume. Also if the underlying block device is lost and came back, the user would still need to remount the disk, since the previous mount point is no longer valid. The only valid use case for remount we thought before is the workload recreated after reboot. Though restarting Docker is similar, which will result in the workload recreation as well. It's maybe hard to tell if the volume engine disappeared due to Docker restart.

Btw, is it what happened in the original case?

**Comment by phal0r** [comment]
It's the same symptoms as in the original case, but the cause is probably not the same. We wait for Rancher 2.2 to have in-depth monitoring information. For now we do not monitor our services and therefor don't know when and why exactly things become unavailable. While we checked another issue we could reproduce the symptoms as described in our original case and wanted to share our findings.

In general I would expect that in failover scenarios the volumes would recover automatically. This is why replicas on different nodes  exist in my understanding. So to understand you correctly, you are anticipating, that on any errors or problems, Kubernetes would remove the pod of a failing workload and create a new one, thus automatically reattaching the volume to the newly generated pod. If a docker service restart triggers the same workflow, the volume should be reattached afterwards to a newly generated workload. We did not check, if the workload actually got recreated, we currently only know  that after the restart the volumes are not available. Based on your explanation, this is not how longhorn should behave. So at least we can reproduce the issue now consistently.

Do you think the problem is related to csi-attacher, csi-provisioner, driver-deployer being also restarted at the same time (since they all run only as one instance, which are located on the node with the restarting docker service) or are we on the wrong track here and the issue comes from somewhere else?

**Comment by phal0r** [comment]
Ok, I just wanted to make sure, we are talking about the same thing. I read in another issue, that you are currently thinking on how to make csi-attacher more fault-tolerant with a raft-like multi-instance setup. So we probably should just wait on how things progress or should we provide more information for this issue?

I really think longhorn is a cool product and it would be nice for outsiders like me to see a kind of roadmap on where you want to head. Does this already exist?

**Comment by yasker** [comment]
@phal0r The high availability with Kubernetes is one important issue we need to address before announcing Longhorn GA. We have already identified a few gaps (including the one you mentioned about Docker and one presuming that Longhorn container killed by Kubernetes or Linux Kernel) and will start working on it next (would be after the holiday though). 

For the time being, I think you can just wait to see if the issue happened again, without trying aggressively to reproduce it. We're planning to release v0.3.3 in January which we hope can include a few node stability fixes. You can see the current (tentative) milestones tracking at https://github.com/rancher/longhorn/milestones.

Thank for providing valuable information to help to improve the product! The storage is hard and we appreciate the feedback very much. Thanks and Happy Holiday!

**Comment by phal0r** [comment]
@yasker 
We updated to 0.3.3 today. Let's how it goes. Deployment was a breeze. Thanks for creating such a nice product. I know storage is hard and there are not to many good projects. We sticked with Longhorn, because we have the impression, that is driven by use cases very close to real environments. Can't wait to see prometheus performance information, volume resizing and more stability updates. Keep up the good work!

**Comment by splattner** [comment]
Is it better with 0.3.3? 
I do also see a lot of randomly failing Volumes (as posten in your Slack Channel).
Adding more Memory helped a bit (as I suspect OOM), but I still see this. Engine Pods are crashing from Time to Time, without any Hint in a Log File. No clue what it is.

**Comment by phal0r** [comment]
Yeah, the errors got less, but still occur. The types of errors we see correspond to #381 . I guess we have to wait until 0.5.0 to see this fixed.

**Comment by yasker** [comment]
@phal0r If the error is caused by network instability or reboot of the node, the fix of #380 may help. It will be included in v0.4.0 release.
